Feature/filetree file status icon

LukasBisz opened the pull request:
## Proposed changes: Colored Status Icons for the File Tree

TLDR: Add colored status icons in file tree

I have added a status Icon feature to the file tree. This feature intruduces small icons next to each file or folder in the tree view, displaying their current status. Each icon is color-coded to provide a quick visual feedback about the status. The colors used for the status icons in the file tree are consistent with the ones used in the badges.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

**Contributor**:
- [ x ] PR is well described and the description can be used as a commit message on squash
- [ x ] Related issues linked to PR if existing and labels set (no issue)
- [ x ] New code is covered with unit tests (no test for frontend available)
- [ ] New/updated ui components are tested inside the storybook (module ui-components only) 
- [ x ] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog`
- [ ] Feature has been tested with different permissions
- [ ] Documentation updated (only necessary for new features or changed behaviour) (docs will be updated once the feature is ready)

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available
- [ ] Check the icons and colors used and provide feedback if adjustments are needed.

fscholdei commented:
Hello LukasBisz,

thank you very much for your contribution! I have spoken to our designer. He also finds the introduction of a status icon in the file tree very helpful. From an accessibility point of view, he considers it very important to display the status using an icon as a further indicator alongside the color. 

However, it is difficult to find descriptive icons and some of them do not provide any additional value. A combined solution for the "Added" and "Deleted" cases using [icon stacking](https://docs.fontawesome.com/web/style/stack) in the existing file icons would be desirable and wouldn't add too much clutter to the interface.

![Design concept](https://github.com/user-attachments/assets/6d16d612-8740-4115-946e-e92d33d89cab)

I also noticed that you format the changeTypes to lower case in [DiffFileTree.tsx#L187](https://github.com/scm-manager/scm-manager/blob/0be63a10aeec5e07b4f089f93ec7d5798a11c26e/scm-ui/ui-components/src/repos/diff/DiffFileTree.tsx#L187), but then check for uppercase in FileChangeTypeIcons. This means that the icons are currently not displayed at all.

Can I assist you further with the implementation?

Regards, Florian

LukasBisz commented:
Hello fscholdei,

Thanks for the feedback! I'll fix the changeTypes formatting mistake to ensure the icons display correctly.
Additionally, I’ll work on implementing the icon stacking approach. I appreciate the offer for assistance! If I run into any issues,
I’ll be sure to reach out.

Regards, Lukas

LukasBisz commented:
Hey,

Sorry about the late reply some things came up that needed my attention. 
I’ve modified the implementation to use icon stacking.
![brave_V1QHHjIBgR](https://github.com/user-attachments/assets/d79294e9-5dda-4bac-ac96-9cab31f6ae4e)

I also added the status icon to the file icon when it's clicked. I would like to hear your opinion on that because i was unsure of usefullness.
![brave_IzfWlUVtZu](https://github.com/user-attachments/assets/70b8fd78-a84d-46af-a574-0f203fa73a3b)

As for the other cases, I’ve decided to keep them in for now. They now use a dot icon, following GitHub's style, except for the "copy" status, which uses the plus icon instead. 
(See here: [GitHub example](https://docs.github.com/assets/cb-16555/images/help/desktop/commit-all.png))
If this is still not wished for i will remove them.

Please let me know if this meets your needs or if any adjustments are required.

Regards, Lukas

fscholdei commented:
Hi Lukas, thanks again for your contribution!

fscholdei commented on scm-ui/ui-components/src/repos/diff/DiffFileTree.tsx:
```diff
@@ -79,6 +82,7 @@ const TreeNode: FC<NodeProps> = ({ node, parentPath, currentFile, setCurrentFile
         </ul>
       ) : (
         <TreeFile
+          changeType={node.changeType}
```
```suggestion
          changeType={node.changeType.toLowerCase() as ChangeType}
```

Since I worked with lowercase below. But I don't think it's very relevant whether this is upper or lower. I just noticed that my suggestion below no longer works otherwise.

fscholdei reviewed (changes requested):
Our designer likes the highlighting of the active file, but would of course like it even more if the next active element was also automatically highlighted when scrolling through the diffs. But we can incorporate this into our backlog for a later expansion phase, also because it could quickly become very complicated.

We came up with the idea of making the active state look similar to the tertiary button. At the beginning, this only has a colored border, in active state it gets solid color, while the content inverts its color to background.

To do this, the type of the outer icon has to change from `far` to `fas` if `isCurrentFile()` and the color of the inner icon from _colorful_ to `has-text-secondary-least`.

LukasBisz commented on scm-ui/ui-components/src/repos/diff/DiffFileTree.tsx:
```diff
@@ -106,13 +202,34 @@ const TreeFile: FC<FileProps> = ({ path, parentPath, currentFile, setCurrentFile
-        <StyledIcon style={{ minWidth: "1.5rem" }} key={completePath + "file"} type="far" alt={t("diff.showContent")}>
-          file
-        </StyledIcon>
+        <span className="fa-stack" style={{ minWidth: "1.5rem", maxWidth: "1.5rem" }}>
+          <StyledIcon
+            className={"fa-stack-2x fa-xs " + getChangeTypeClassName(changeType)}
+            style={{ minWidth: "1.5rem", fontSize: "1.5rem" }}
+            key={completePath + "file"}
+            type="far"
+            alt={t("diff.showContent")}
+          >
+            file
+          </StyledIcon>
+          <span className="fa-stack-1x fa-xs" style={{ minWidth: "1.5rem", maxWidth: "1.5rem" }}>
+            <FileChangeTypeIcon changeType={changeType.toUpperCase()} />
+          </span>
+        </span>
```
Hi Florian,

I implemented the changes you suggested but noticed an problem.
When I remove the custom styling, all the icons become larger, and the file names overlap with them.
![brave_hFE3GN4Hfl](https://github.com/user-attachments/assets/65b8b160-e33b-420f-9963-78e241b89908)

Am i missing something, or is there another way to approach this?

Thanks, 
Lukas

fscholdei reviewed (commented):
Does that still correspond to your idea? Can you follow my changes or would you like to adjust something?

fscholdei commented:
I'll have another dev look at this feature for a review. Then it should hopefully be part of an upcoming release.

pfeuffer commented:
Thanks for your support, @LukasBisz 💙
